org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/markers/ApplyQuickFixMarkerResolver.java

```diff
@@ -0,0 +1,99 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarlint.eclipse.ui.internal.markers;
+
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.ui.IEditorPart;
+import org.eclipse.ui.IMarkerResolution2;
+import org.eclipse.ui.texteditor.ITextEditor;
+import org.sonarlint.eclipse.core.SonarLintLogger;
+import org.sonarlint.eclipse.core.internal.adapter.Adapters;
+import org.sonarlint.eclipse.core.internal.quickfixes.MarkerQuickFix;
+import org.sonarlint.eclipse.core.resource.ISonarLintFile;
+import org.sonarlint.eclipse.ui.internal.SonarLintImages;
+import org.sonarlint.eclipse.ui.internal.util.PlatformUtils;
+
+public class ApplyQuickFixMarkerResolver implements IMarkerResolution2 {
+
+  private final MarkerQuickFix fix;
+
+  public ApplyQuickFixMarkerResolver(MarkerQuickFix fix) {
+    this.fix = fix;
+  }
+
+  @Override
+  public String getDescription() {
+    return "Automatically modifies the code to fix the issue";
+  }
+
+  @Override
+  public String getLabel() {
+    return fix.getMessage();
+  }
+
+  @Override
+  public void run(IMarker marker) {
+    ISonarLintFile file = getFile(marker);
+    if (file == null) {
+      return;
+    }
+    IDocument document = getDocument(file);
+    Display.getDefault().asyncExec(() -> fix.getTextEdits().forEach(textEdit -> {
+      try {
+        IMarker editMarker = textEdit.getMarker();
+        int startOffset = (int) editMarker.getAttribute(IMarker.CHAR_START);
+        int endOffset = (int) editMarker.getAttribute(IMarker.CHAR_END);
+        document.replace(startOffset, endOffset - startOffset, textEdit.getNewText());
+      } catch (Exception e) {
+        SonarLintLogger.get().error("Quick fix location does not exist", e);
+      }
+    }));
+  }
+
+  @Nullable
+  private static ISonarLintFile getFile(IMarker marker) {
+    IFile file = Adapters.adapt(marker.getResource(), IFile.class);
+    if (file != null) {
+      return Adapters.adapt(file, ISonarLintFile.class);
+    }
+    return null;
+  }
+
+  private static IDocument getDocument(ISonarLintFile file) {
+    IDocument doc;
+    IEditorPart editorPart = PlatformUtils.findEditor(file);
+    if (editorPart instanceof ITextEditor) {
+      doc = ((ITextEditor) editorPart).getDocumentProvider().getDocument(editorPart.getEditorInput());
+    } else {
+      doc = file.getDocument();
```

Comment:
This part doesn't work. If the file is not open in an editor, this code will return an "in memory" IDocument created from the file content. Changes to this document will have no effect on the file itself.

Reply:
Is it what you refered to in your list as "support applying quick fixes on closed files" ? If so I think I will merge this one and I will add a separate ticket.

Reply:
yes